### PR TITLE
Add MiniAudio Gem from https://github.com/AMZN-Olex/MiniAudioO3DE

### PR DESCRIPTION
**Pull request by AMZN-alexpete**
## What does this PR do?

1. Copies MiniAudioO3DE from https://github.com/AMZN-Olex/MiniAudioO3DE  https://github.com/AMZN-Olex/MiniAudioO3DE/commit/ef13936420c191c0b0220f09cc7e2e867896fabe into the `Gems` folder.
2. Updates `engine.json` to add the `external_subdirectory`

NOTE: This PR does not make the Gem compatible with the ATL and Audio Controls Editor systems like Wwise, but that should be done in a future PR.

## How was this PR tested?

- Built a project in profile and release with the gem activated
- Tested that I could play sounds in the Editor using the MiniAudio playback component.

### CODE REVIEW

**Comment by gralco** [comment]
Hi, your README says the following:

> Supported Sound Formats
> 
> *    .wav
> *    .ogg
> 
> MiniAudio supports more formats, but not all are supported in this Gem yet.

But back when I tested [AMZN-Olex/MiniAudioO3DE](https://github.com/AMZN-Olex/MiniAudioO3DE) I tried mp3, flac, wav, and ogg, all of which worked for me.

According to the MiniAudio documentation you'd have to explicitly disable flac and mp3 support, was this done, if so why?

https://miniaud.io/docs/manual/index.html#Decoding

**Comment by nick-l-o3de** [comment]
> Hi, your README says the following:
> 
> > Supported Sound Formats
> > 
> > * .wav
> > * .ogg
> > 
> > MiniAudio supports more formats, but not all are supported in this Gem yet.
> 
> But back when I tested [AMZN-Olex/MiniAudioO3DE](https://github.com/AMZN-Olex/MiniAudioO3DE) I tried mp3, flac, wav, and ogg, all of which worked for me.
> 
> According to the MiniAudio documentation you'd have to explicitly disable flac and mp3 support, was this done, if so why?
> 
> https://miniaud.io/docs/manual/index.html#Decoding

I don't see anywhere in the commit where those systems are explicitly disabled, its using the defaults, so I have to assume they are enabled.  that does raise a question though - some of those formats are patent encumbered and may not be suitable to ship in a free, open source ecosystem.

File: Gems/MiniAudio/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+set(gem_path ${CMAKE_CURRENT_LIST_DIR})
+set(gem_json ${gem_path}/gem.json)
+o3de_restricted_path(${gem_json} gem_restricted_path gem_parent_relative_path)
+
+o3de_pal_dir(pal_dir ${CMAKE_CURRENT_LIST_DIR}/Platform/${PAL_PLATFORM_NAME} "${gem_restricted_path}" "${gem_path}" "${gem_parent_relative_path}")
```
Comment (lumberyard-employee-dm) [inline Gems/MiniAudio/CMakeLists.txt:13]: Can you update this block to provide a call to [o3de_gem_setup("MiniAudio")](https://github.com/o3de/o3de/blob/development/Templates/DefaultGem/Template/CMakeLists.txt#L9).

File: Gems/MiniAudio/Code/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+o3de_pal_dir(pal_dir ${CMAKE_CURRENT_LIST_DIR}/Platform/${PAL_PLATFORM_NAME} "${gem_restricted_path}" "${gem_path}" "${gem_parent_relative_path}")
+
+# Now that we have the platform abstraction layer (PAL) folder for this folder, thats where we will find the
+# traits for this platform. Traits for a platform are defines for things like whether or not something in this gem
+# is supported by this platform.
+include(${pal_dir}/PAL_${PAL_PLATFORM_NAME_LOWERCASE}.cmake)
+
+# The MiniAudio.API target declares the common interface that users of this gem should depend on in their targets
+ly_add_target(
+    NAME MiniAudio.API INTERFACE
```
Comment (lumberyard-employee-dm) [inline Gems/MiniAudio/Code/CMakeLists.txt:24]: The MiniAudio targets can be substituted with `${gem_name}` to make it easier to rename later. 
This needs to be combined with the `o3de_gem_setup` call from the previous commen.t

File: Gems/MiniAudio/Code/Source/Clients/MiniAudioModule.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    class MiniAudioModule
+        : public MiniAudioModuleInterface
+    {
+    public:
+        AZ_RTTI(MiniAudioModule, "{501C94A1-993A-4203-9720-D43D6C1DDB7A}", MiniAudioModuleInterface);
+        AZ_CLASS_ALLOCATOR(MiniAudioModule, AZ::SystemAllocator, 0);
+    };
+}// namespace MiniAudio
+
+AZ_DECLARE_MODULE_CLASS(Gem_MiniAudio, MiniAudio::MiniAudioModule)
```
Comment (lumberyard-employee-dm) [inline Gems/MiniAudio/Code/Source/Clients/MiniAudioModule.cpp:23]: The AZ_DECLARE_MODULE_CLASS macro can be updated to make it simpler to rename the gem by updating the gem.json by replicating the changes from the Gem templates [${Name}Module.cpp](https://github.com/o3de/o3de/blob/development/Templates/DefaultGem/Template/Code/Source/Clients/%24%7BName%7DModule.cpp#L26-L30) file
```suggestion
#if defined(O3DE_GEM_NAME)
AZ_DECLARE_MODULE_CLASS(AZ_JOIN(Gem_, O3DE_GEM_NAME), MiniAudio::MiniAudioModule)
#else
AZ_DECLARE_MODULE_CLASS(Gem_MiniAudio, MiniAudio::MiniAudioModule)
#endif
```

File: Gems/MiniAudio/Code/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+# By default, we will specify that the above target MiniAudio would be used by
+# Client and Server type targets when this gem is enabled.  If you don't want it
+# active in Clients or Servers by default, delete one of both of the following lines:
+ly_create_alias(NAME MiniAudio.Clients NAMESPACE Gem TARGETS Gem::MiniAudio)
+ly_create_alias(NAME MiniAudio.Servers NAMESPACE Gem TARGETS Gem::MiniAudio)
+
+# For the Client and Server variants of MiniAudio Gem, an alias to the MiniAudio.API target will be made
+ly_create_alias(NAME MiniAudio.Clients.API NAMESPACE Gem TARGETS Gem::MiniAudio.API)
+ly_create_alias(NAME MiniAudio.Servers.API NAMESPACE Gem TARGETS Gem::MiniAudio.API)
```
Comment (lumberyard-employee-dm) [inline Gems/MiniAudio/Code/CMakeLists.txt:87]: Adding a source file propertie to inject the O3DE_GEM_NAME define into the MiniAudioModule.cpp will complete the work needed to simplify renaming the gem in the future if need be
https://github.com/o3de/o3de/blob/development/Templates/DefaultGem/Template/Code/CMakeLists.txt#L81-L92

**Comment by AMZN-alexpete** [comment]
> > Hi, your README says the following:
> > > Supported Sound Formats
> > > 
> > > * .wav
> > > * .ogg
> > > 
> > > MiniAudio supports more formats, but not all are supported in this Gem yet.
> > 
> > 
> > But back when I tested [AMZN-Olex/MiniAudioO3DE](https://github.com/AMZN-Olex/MiniAudioO3DE) I tried mp3, flac, wav, and ogg, all of which worked for me.
> > According to the MiniAudio documentation you'd have to explicitly disable flac and mp3 support, was this done, if so why?
> > https://miniaud.io/docs/manual/index.html#Decoding
> 
> I don't see anywhere in the commit where those systems are explicitly disabled, its using the defaults, so I have to assume they are enabled. that does raise a question though - some of those formats are patent encumbered and may not be suitable to ship in a free, open source ecosystem.
> 
> Edit: MP3 decoding appears to be patent free even in the US. So maybe its safe...?

I'll add the missing formats (mp3 and flac)

[Flac claims to be the only non-proprietary and unencumbered by patents, with open-source implementation format ](https://xiph.org/flac/features.html)

Supposedly [MP3 patents have expired](https://en.wikipedia.org/wiki/MP3#:~:text=The%20basic%20MP3%20decoding%20and,US%20between%202007%20and%202017.), and other open source engines [such as Godot allow mp3 playback](https://docs.godotengine.org/en/stable/tutorials/assets_pipeline/importing_audio_samples.html) also

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**